src/adaptation/precond.jl

```diff
@@ -2,389 +2,211 @@
-    @unpack μ, M, n = wv
-    for i in eachindex(s)
-        δ = s[i] - μ[i]
-        μ[i] += δ / n
-        M[i] += δ * (s[i] - μ[i])
-    end
-end
+getM⁻¹(::UnitMassMatrix{T}) where {T} = LinearAlgebra.UniformScaling{T}(one(T))
 
-# https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/var_adaptation.hpp
-function get_var(wv::WelfordVar{<:AbstractVecOrMat{T}}) where {T<:AbstractFloat}
-    n, M = T(wv.n), wv.M
-    @assert n >= 2 "Cannot get covariance with only one sample"
-    return (n / ((n + 5) * (n - 1))) .* M .+ T(1e-3) * (5 / (n + 5))
-end
+adapt!(
+    ::UnitMassMatrix,
+    ::AbstractVecOrMat{<:AbstractFloat},
+    ::AbstractScalarOrVec{<:AbstractFloat},
+    is_update::Bool=true
+) = nothing
 
-## Covariance estimator
+## Diagonal mass matrix adaptor
 
-abstract type CovEstimator{T} end
+abstract type VarEstimator{T} <: MassMatrixAdaptor end
 
-# NOTE: This naive covariance estimator is used only in testing.
-mutable struct NaiveCov{T<:AbstractVector{<:AbstractVector{<:AbstractFloat}}} <: CovEstimator{T}
-    n :: Int
-    S :: T
-end
+getM⁻¹(ve::VarEstimator) = ve.var
 
-NaiveCov(::Type{T}, sz::Union{Tuple{Int}}) where {T<:AbstractFloat} = NaiveCov(0, Vector{Vector{T}}())
+Base.string(ve::VarEstimator) = string(getM⁻¹(ve))
 
-function add_sample!(nc::NaiveCov, s::AbstractVector)
-    nc.n += 1
-    push!(nc.S, s)
+function update!(ve::VarEstimator)
+    ve.n >= ve.n_min && (ve.var .= getest(ve))
 end
 
-function get_cov(nc::NaiveCov)
-    @assert nc.n >= 2 "Cannot get covariance with only one sample"
-    return Statistics.cov(nc.S)
+# NOTE: this naive variance estimator is used only in testing
+struct NaiveVar{T<:AbstractFloat, E<:AbstractVector{<:AbstractVecOrMat{T}}} <: VarEstimator{T}
+    S :: E
+    NaiveVar(S::E) where {E} = new{eltype(eltype(E)), E}(S)
 end
 
-function reset!(nc::NaiveCov{T}) where {T}
-    nc.n = 0
-    nc.S = T()
-end
+NaiveVar{T}(sz::Tuple{Int}) where {T<:AbstractFloat} = NaiveVar(Vector{Vector{T}}())
+NaiveVar{T}(sz::Tuple{Int,Int}) where {T<:AbstractFloat} = NaiveVar(Vector{Matrix{T}}())
 
-# Ref: https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_covar_estimator.hpp
-mutable struct WelfordCov{T<:AbstractFloat} <: CovEstimator{T}
-    n :: Int
-    μ :: Vector{T}
-    M :: Matrix{T}
-    δ :: Vector{T} # temporary
-end
+NaiveVar(sz::Union{Tuple{Int}, Tuple{Int,Int}}) = NaiveVar{Float64}(sz)
 
-function WelfordCov(::Type{T}, sz::Tuple{Int}) where {T<:AbstractFloat}
-    d, = sz
-    WelfordCov(0, zeros(T, d), zeros(T, d, d), zeros(T, d))
-end
+Base.push!(nv::NaiveVar, s::AbstractVecOrMat) = push!(nv.S, s)
 
-function reset!(wc::WelfordCov{T}) where {T<:AbstractFloat}
-    wc.n = 0
-    wc.μ .= zero(T)
-    wc.M .= zero(T)
-end
+reset!(nv::NaiveVar) = resize!(nv.S, 0)
 
-function add_sample!(wc::WelfordCov, s::AbstractVector)
-    wc.n += 1
-    @unpack δ, μ, n, M = wc
-    δ .= s .- μ
-    μ .+= δ ./ n
-    M .+= (s .- μ) .* δ'
+function getest(nv::NaiveVar)
```

Comment:
`getest` can be easily confused with `get_test`, unfortunately. Why not keep `get_var` and `get_cov` instead?

Reply:
Just for consistency. Calling `getest(::VarEstimator)` or `getest(::CovEstimator)` is self-explianed on what is expected to return. If the issues is about confusing against `get_test`, we can do `get_estimation`. I don't mind revert to the old names as well.

Reply:
`get_estimation` sounds good.